Add a basic glossary to the docs (#104)
XanaduAI/strawberryfields: doc/glossary.rst

Glossary
========

Expressions that are grouped together are used synonymously in the Strawberry Fields documentation.

.. glossary::

   backend
      Executes quantum :term:`programs <program>`. Can be either a classical simulator or a
      quantum hardware device, local or remote.
      See: :class:`.BaseBackend`

   circuit
   quantum circuit
   program
      A sequence of quantum :term:`operations <operation>` acting on a quantum :term:`register`.
      Typically contains :term:`measurements <measurement>` that produce a stochastic classical output.
      Represents a quantum computation.
      See: :class:`.Program`

   compilation
      The process of converting a quantum :term:`program` into an :term:`equivalent program` that
      can be run on a specific :term:`device`. The compilation can fail if the device is incapable
      of running the program.

>    device

Reviewer: @smite @josh146 the name "device" is definitely the most unclear concept. Suggest choosing better name

Possibilities:
1. circuit family
1. circuit type
1. circuit class
1. program spec
1. subprogram
1. program family
1. program type
1. program class

Reviewer 2: Of the above, I think I like 'circuit class' or 'circuit type' the most, but it still doesn't sound quite right...

Author: I guess "circuit template" is already taken, and means a parametrized circuit?

      Something that can run a well-defined subclass of quantum :term:`programs <program>`.

>       Each :term:`backend` corresponds to one or more devices.

Reviewer: ```suggestion
      The same device can potentially be evaluated on more than one :term:`backend`, and each :term:`backend` can implement one or more devices.
```

      See: :class:`.DeviceSpecs`

   equivalent circuit
   equivalent program
      Two quantum :term:`circuits <circuit>` are equivalent iff they produce the same output
      probability distributions.

   engine

>    frontend

Reviewer: ```suggestion
```

      A class for executing quantum :term:`programs <program>` on a specific :term:`backend`.

>       See: :class:`.BaseEngine`

Author: Yeah, I guess so. Let's remove frontend from the glossary for now.

>       See: :class:`.BaseEngine`

Reviewer: no need to remove, let's just add another glossary term, with content similar to what @josh146 wrote above

>       See: :class:`.BaseEngine`

Reviewer: done

   gate
   quantum gate
      A quantum :term:`operation` that changes the state of the :term:`register`.
      Analogous to a logic gate in a Boolean circuit.

   measurement
      A quantum :term:`operation` that returns classical information.

   operation
   quantum operation
      Something that acts on specific subsystems of the :term:`register`, changing its state and

>       possibly returning classical information, e.g. a state preparation, :term:`gate`,

Reviewer: ```suggestion
      possibly returning classical information, e.g., a state preparation, :term:`gate`,
```

>       :term:`measurement` etc.

Reviewer: ```suggestion
      :term:`measurement`, etc.
```

      See: :class:`.Operation`

   register

>    quantum register

Reviewer 2: Should we add 'subsystems' and 'modes' here as well --- we tend to use all four throughout the documentation in different places

Author: right

>    quantum register

Reviewer: ```suggestion
   quantum register
   system
   mode
   qumode
```

      A quantum system (or a simulation of one) storing the state of the quantum computation.

>       Typically consists of several subsystems (which in CV quantum computing are called modes).

Reviewer: could add `state preparation`, but not urgent
